File: placement-flutter/storyly_placement_flutter/ios/Classes/Common/SPPlacementProviderManager.swift
import Foundation
import StorylyPlacement
import StorylyCore

@objc public class SPPlacementProviderManager: NSObject {
    
    @objc public static let shared = SPPlacementProviderManager()
    
    private var providers: [String: SPPlacementProviderWrapper] = [:]

    private override init() {
        super.init()
    }
    
    @objc public func createProvider(id: String) -> SPPlacementProviderWrapper {
        print("[SPPlacementProviderManager] Create provider: \(id)")
        let wrapper = SPPlacementProviderWrapper(id: id)
        providers[id] = wrapper
        return wrapper
    }
    
    @objc public func getProvider(id: String) -> SPPlacementProviderWrapper? {
        return providers[id]
    }
    
    @objc public func destroyProvider(id: String) {
        print("[SPPlacementProviderManager] Destroy provider: \(id)")
        providers.removeValue(forKey: id)
    }
}

@objc public class SPPlacementProviderWrapper: NSObject {
    
    @objc public let id: String
    @objc public lazy var provider: STRPlacementDataProvider = {
        return STRPlacementDataProvider()
    }()
    
    @objc public var sendEvent: ((String, SPPlacementProviderEventType, String) -> Void)?
    
    init(id: String) {
        self.id = id
        super.init()
    }
    
    @objc public func configure(configJson: String) {
        DispatchQueue.main.async {
          guard let parsedConfig = decodeFromJson(configJson) else {
              print("[SPPlacementProviderWrapper] Failed to parse config JSON")
              return
          }
          
          self.setupProvider(config: parsedConfig)
        }
    }
    
    private func setupProvider(config: [String: Any]) {
        DispatchQueue.main.async {
            guard let token = config["token"] as? String else {
              print("[SPPlacementProviderWrapper] Token not found in config")
              return
            }
            
            print("[SPPlacementProviderWrapper] Configuring provider with token: \(token)")
            
            let placementConfig = decodeSTRPlacementConfig(config, token: token)
            
            self.provider.delegate = STRProviderDelegateImpl(wrapper: self)
            self.provider.productDelegate = STRProviderProductDelegateImpl(wrapper: self)
            self.provider.config = placementConfig
        }
    }
    
    @objc public func hydrateProducts(productsJson: String) {
        DispatchQueue.main.async {
            guard let dict = decodeFromJson(productsJson),
                  let productsArray = dict["products"] as? [[String: Any]] else {
                return
            }
            
            print("[SPPlacementProviderWrapper] hydrateProducts: \(productsJson)")
            
            let products = productsArray.compactMap { decodeSTRProductItem($0) }
            self.provider.hydrateProducts(products: products)
        }
    }
    
    @objc public func hydrateWishlist(productsJson: String) {
        DispatchQueue.main.async {
            guard let dict = decodeFromJson(productsJson),
                  let productsArray = dict["products"] as? [[String: Any]] else {
                return
            }
            
            print("[SPPlacementProviderWrapper] hydrateWishlist: \(productsJson)")
            
            let products = productsArray.compactMap { decodeSTRProductItem($0) }
            self.provider.hydrateWishlist(products: products)
        }
    }
}

// MARK: - STRDataProviderListener Implementation

private class STRProviderDelegateImpl: NSObject, STRDataProviderDelegate {
    weak var wrapper: SPPlacementProviderWrapper?
    
    init(wrapper: SPPlacementProviderWrapper) {
        self.wrapper = wrapper
    }
    
    func onLoad(data: STRDataPayload, dataSource: STRDataSource) {
        guard let wrapper = wrapper else { return }
        
        let eventData: [String: Any] = [
            "data": encodeDataPayload(data),
            "dataSource": dataSource.description
        ]
        
        if let eventJson = encodeToJson(eventData) {
            print("[SPPlacementProviderWrapper] STRDataProviderListener:onLoad: \(eventJson)")
            wrapper.sendEvent?(wrapper.id, .onLoad, eventJson)
        }
    }
    
    func onLoadFail(errorMessage: String) {
        guard let wrapper = wrapper else { return }
        
        let eventData: [String: Any] = [
            "errorMessage": errorMessage
        ]
        
        if let eventJson = encodeToJson(eventData) {
            print("[SPPlacementProviderWrapper] STRDataProviderListener:onLoadFail: \(eventJson)")
          wrapper.sendEvent?(wrapper.id, .onLoadFail, eventJson)
        }
    }
}

// MARK: - STRDataProviderProductListener Implementation

private class STRProviderProductDelegateImpl: NSObject, STRDataProviderProductDelegate {
    weak var wrapper: SPPlacementProviderWrapper?
    
    init(wrapper: SPPlacementProviderWrapper) {
        self.wrapper = wrapper
    }
    
    func onHydration(products: [STRProductInformation]) {
        guard let wrapper = wrapper else { return }
        
        let eventData: [String: Any] = [
            "products": products.map { encodeSTRProductInformation($0) }
        ]
        
        if let eventJson = encodeToJson(eventData) {
            print("[SPPlacementProviderWrapper] STRDataProviderProductListener:onHydration: \(eventJson)")
            wrapper.sendEvent?(wrapper.id, .onHydration, eventJson)
        }
    }
}

File: flutter/storyly_placement_flutter/android/src/main/kotlin/com/appsamurai/storyly/storyly_placement_flutter/common/SPPlacementProviderManager.kt
package com.appsamurai.storyly.storyly_placement_flutter.common

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.appsamurai.storyly.core.data.model.STRDataPayload
import com.appsamurai.storyly.core.data.model.STRDataSource
import com.appsamurai.storyly.core.data.model.product.STRProductInformation
import com.appsamurai.storyly.core.listener.log.STRLog
import com.appsamurai.storyly.core.listener.provider.STRDataProviderListener
import com.appsamurai.storyly.core.listener.provider.STRDataProviderProductListener
import com.appsamurai.storyly.placement.data.provider.STRPlacementDataProvider
import com.appsamurai.storyly.storyly_placement_flutter.common.data.decodeSTRPlacementConfig
import com.appsamurai.storyly.storyly_placement_flutter.common.data.encodeDataPayload
import com.appsamurai.storyly.storyly_placement_flutter.common.data.product.decodeSTRProductItem
import com.appsamurai.storyly.storyly_placement_flutter.common.data.product.decodeSTRProductInformation
import com.appsamurai.storyly.storyly_placement_flutter.common.data.product.encodeSTRProductInformation
import com.appsamurai.storyly.storyly_placement_flutter.common.data.product.encodeSTRProductItem
import com.appsamurai.storyly.storyly_placement_flutter.common.data.util.decodeFromJson
import com.appsamurai.storyly.storyly_placement_flutter.common.data.util.encodeToJson


object SPPlacementProviderManager {
    private val providers = mutableMapOf<String, SPPlacementProviderWrapper>()
    private val lock = Any()

    fun createProvider(context: Context, id: String): SPPlacementProviderWrapper {
        synchronized(lock) {
            val wrapper = SPPlacementProviderWrapper(context, id)
            providers[id] = wrapper
            return wrapper
        }
    }

    fun getProvider(id: String): SPPlacementProviderWrapper? {
        synchronized(lock) {
            return providers[id]
        }
    }

    fun destroyProvider(id: String) {
        synchronized(lock) {
            providers.remove(id)
        }
    }
}

class SPPlacementProviderWrapper(
    private val context: Context,
    val id: String
) {
    val provider: STRPlacementDataProvider by lazy { STRPlacementDataProvider(context) }

    internal var sendEvent: ((String, SPPlacementProviderEventType, String) -> Unit)? = null

    fun configure(configJson: String) {
        Handler(Looper.getMainLooper()).post {
            val parsedConfig = decodeFromJson(configJson) ?: run {
                STRLog.error("[SPPlacementProviderWrapper] Failed to parse config JSON")
                return@post
            }

            setupProvider(parsedConfig)
        }
    }

    private fun setupProvider(config: Map<String, Any?>) {
        val token = config["token"] as? String ?: return


        val placementConfig = decodeSTRPlacementConfig(config, token)
        placementConfig.framework = "flutter"

        provider.apply {
          listener = object : STRDataProviderListener {
            override fun onLoad(data: STRDataPayload, dataSource: STRDataSource) {
              val eventJson = encodeToJson(mapOf(
                  "data" to encodeDataPayload(data),
                  "dataSource" to dataSource.value,
              ))
              sendEvent?.invoke(id, SPPlacementProviderEventType.ON_LOAD, eventJson ?: "")
            }

            override fun onLoadFail(errorMessage: String) {
              val eventJson = encodeToJson(mapOf(
                "errorMessage" to errorMessage,
              ))
              sendEvent?.invoke(id, SPPlacementProviderEventType.ON_LOAD_FAIL, eventJson ?: "")
            }
          }
          productListener = object : STRDataProviderProductListener {
            override fun onHydration(products: List<STRProductInformation>) {
              val eventJson = encodeToJson(mapOf(
                "products" to products.map { encodeSTRProductInformation(it) },
              ))
              sendEvent?.invoke(id, SPPlacementProviderEventType.ON_HYDRATION, eventJson ?: "")
            }
          }

          this.config = placementConfig
        }
    }

    fun hydrateProducts(raw: String) {
        Handler(context.mainLooper).post {
            val map = decodeFromJson(raw) ?: return@post
            val products = (map["products"] as? List<Map<String, Any?>>)?.mapNotNull {
                decodeSTRProductItem(it)
            } ?: return@post
            provider.hydrateProducts(products)
        }
    }

    fun hydrateWishlist(raw: String) {
        Handler(context.mainLooper).post {
            val map = decodeFromJson(raw) ?: return@post
            val products = (map["products"] as? List<Map<String, Any?>>)?.mapNotNull {
                decodeSTRProductInformation(it)
            } ?: return@post
            provider.hydrateWishlist(products)
        }
    }
}
